src/router/screen/doc/screen.1
.\" vi:set wm=5
.TH SCREEN 1 "May 2025" "GNU Screen 5.0.1"
.
.
.SH NAME
screen \- screen manager with VT100/ANSI terminal emulation


.SH SYNOPSIS
.B screen
[
.B \-\fIoptions\fP
] [
.B \fIcmd\fP
[
.B \fIargs\fP
] ]
.br
.B screen \-r
[[\fIpid\fP\fB.\fP]\fItty\fP[\fB.\fP\fIhost\fP]]
.br
.B screen \-r
\fIsessionowner\fP\fB/\fP[[\fIpid\fP\fB.\fP]\fItty\fP[\fB.\fP\fIhost\fP]]
.ta .5i 1.8i


.SH DESCRIPTION
.I Screen
is a full-screen window manager that
multiplexes a physical terminal between several processes (typically
interactive shells).
Each virtual terminal provides the functions
of a DEC VT100 terminal and, in addition, several control functions
from the ISO 6429 (ECMA 48, ANSI X3.64) and ISO 2022 standards
(e.\|g. insert/delete line and support for multiple character sets).
There is a scrollback history buffer for each virtual terminal and a
copy-and-paste mechanism that allows moving text regions between
windows.
.PP
When
.I screen
is called, it creates a single window with a shell in it (or the specified
command) and then gets out of your way so that you can use the program as you
normally would.
Then, at any time, you can create new (full-screen) windows with other programs
in them (including more shells), kill existing windows, view a list of
windows, turn output logging on and off, copy-and-paste text between
windows, view the scrollback history, switch between windows
in whatever manner you wish, etc. All windows run their programs completely
independent of each other. Programs continue to run when their window
is currently not visible and even when the whole 
.I screen 
session is detached from the user's terminal.  When a program terminates,
.I screen
(per default) kills the window that contained it.
If this window was in the foreground, the display switches to the previous
window; if none are left,
.I screen
exits. Shells usually distinguish between running as login-shell or sub-shell.
Screen runs them as sub-shells, unless told otherwise (See \*Qshell\*U .screenrc command).
.PP
Everything you type is sent to the program running in the current window.
The only exception to this is the one keystroke that is used to initiate
a command to the window manager.
By default, each command begins with a control-a (abbreviated C-a from
now on), and is followed by one other keystroke.
The command character and all the key bindings can be fully customized
to be anything you like, though they are always two characters in length.
.PP
.I Screen
does not understand the prefix \*QC-\*U to mean control, although this notation is 
used in this manual for readability. 
Please use the caret notation (\*Q^A\*U instead of \*QC-a\*U) as arguments
to e.g. the 
.I escape
command or the \fI-e\fP option.
.I Screen
will also print out control characters in caret notation.
.PP
The standard way to create a new window is to type \fBC-a c\fP.
This creates a new window running a shell and switches to that
window immediately, regardless of the state of the process running
in the current window.
Similarly, you can create a new window with a custom command in it by
first binding the command to a keystroke (in your .screenrc file or at the
\fBC-a :\fP command line) and
then using it just like the \fBC-a c\fP command.
In addition, new windows can be created by running a command like:
.IP
screen emacs prog.c
.PP
from a shell prompt within a previously created window.
This will not run another copy of
.IR screen ,
but will instead supply the command name and its arguments to the window
manager (specified in the $STY environment variable) who will use it to
create the new window.
The above example would start the emacs editor (editing prog.c) and switch
to its window. - Note that you cannot transport environment variables from
the invoking shell to the application (emacs in this case), because it is
forked from the parent screen process, not from the invoking shell.
.PP
If \*Q/etc/utmp\*U is writable by
.IR screen ,
an appropriate record will be written to this file for each window, and
removed when the window is terminated.
This is useful for working with \*Qtalk\*U, \*Qscript\*U, \*Qshutdown\*U,
\*Qrsend\*U, \*Qsccs\*U and other similar programs that use the utmp
file to determine who you are. As long as
.I screen
is active on your terminal,
the terminal's own record is removed from the utmp file. See also \*QC-a L\*U.


.SH GETTING STARTED
Before you begin to use
.I screen
you'll need to make sure you have correctly selected your terminal type,
just as you would for any other termcap/terminfo program.
(You can do this by using
.IR test
for example.)
.PP
If you're impatient and want to get started without doing a lot more reading,
you should remember this one command:  \fBC-a ?\fP.
Typing these two characters will display a list of the available
.I screen
commands and their bindings. Each keystroke is discussed in
the section \*QDEFAULT KEY BINDINGS\*U. The manual section \*QCUSTOMIZATION\*U
deals with the contents of your .screenrc.
.PP
If your terminal is a \*Qtrue\*U auto-margin terminal (it doesn't allow
the last position on the screen to be updated without scrolling the
screen) consider using a version of your terminal's termcap that has
automatic margins turned \fIoff\fP. This will ensure an accurate and
optimal update of the screen in all circumstances. Most terminals
nowadays have \*Qmagic\*U margins (automatic margins plus usable last
column). This is the VT100 style type and perfectly suited for
.IR screen .
If all you've got is a \*Qtrue\*U auto-margin terminal 
.I screen
will be content to use it, but updating a character put into the last
position on the screen may not be possible until the screen scrolls or
the character is moved into a safe position in some other way. This
delay can be shortened by using a terminal with insert-character
capability.


.SH "COMMAND-LINE OPTIONS"
Screen has the following command-line options:
.TP 5
.B \-a
include \fIall\fP capabilities (with some minor exceptions) in each
window's termcap, even if
.I screen
must redraw parts of the display in order to implement a function.
.TP 5
.B \-A
Adapt the sizes of all windows to the size of the current terminal.
By default,
.I screen
tries to restore its old window sizes when attaching to resizable terminals
(those with \*QWS\*U in its description, e.g. suncmd or some xterm).
.TP 5
.BI "\-c " file
override the default configuration file from \*Q$HOME/.screenrc\*U
to \fIfile\fP.
.TP 5
.BR \-d | \-D " [" \fIpid.tty.host\fR ]
does not start
.IR screen ,
but detaches the elsewhere running
.I screen
session. It has the same effect as typing \*QC-a d\*U from
.IR screen 's
controlling terminal. \fB\-D\fP is the equivalent to the power detach key.
If no session can be detached, this option is ignored. In combination with the
\fB\-r\fP/\fB\-R\fP option more powerful effects can be achieved:
.TP 8
.B \-d \-r
Reattach a session and if necessary detach it first.
.TP 8
.B \-d \-R
Reattach a session and if necessary detach or even create it first.
.TP 8
.B \-d \-RR
Reattach a session and if necessary detach or create it. Use the first
session if more than one session is available.
.TP 8
.B \-D \-r
Reattach a session. If necessary detach and logout remotely first.
.TP 8
.B \-D \-R
Attach here and now. In detail this means: If a session is running, then
reattach. If necessary detach and logout remotely first.
If it was not running create it and notify the user. This is the
author's favorite.
.TP 8
.B \-D \-RR
Attach here and now. Whatever that means, just do it.
.IP "" 5
Note: It is always a good idea to check the status of your sessions by means of
\*Qscreen \-list\*U.
.TP 5
.BI "\-e " xy
specifies the command character to be \fIx\fP and the character generating a
literal command character to \fIy\fP (when typed after the command character).
The default is \*QC-a\*U and `a', which can be specified as \*Q-e^Aa\*U.
When creating a
.I screen
session, this option sets the default command character. In a multiuser
session all users added will start off with this command character. But
when attaching to an already running session, this option changes only
the command character of the attaching user.
This option is equivalent to either the commands \*Qdefescape\*U or
\*Qescape\*U respectively.
.TP 5
.BR \-f\fP ", " \-fn ", and " \-fa
turns flow-control on, off, or \*Qautomatic switching mode\*U.
This can also be defined through the \*Qdefflow\*U .screenrc command.
.TP 5
.BI "\-h " num
Specifies the history scrollback buffer to be \fInum\fP lines high.
.TP 5
.B \-i
will cause the interrupt key (usually C-c) to interrupt the display
immediately when flow-control is on.
See the \*Qdefflow\*U .screenrc command for details.
The use of this option is discouraged.
.TP 5
.BR \-l " and " \-ln
turns login mode on or off (for /etc/utmp updating).
This can also be defined through the \*Qdeflogin\*U .screenrc command.
.TP 5
.BR \-ls " [" \fImatch ]
.PD 0
.TP 5
.BR \-list " [" \fImatch ] 
.PD
does not start
.IR screen ,
but prints a list of
.I pid.tty.host
strings identifying your
.I screen
sessions.
Sessions marked `detached' can be resumed with \*Qscreen \-r\*U. Those marked
`attached' are running and have a controlling terminal. If the session runs in 
multiuser mode, it is marked `multi'. Sessions marked as `unreachable' either
live on a different host or are `dead'.
An unreachable session is considered dead, when its name
matches either the name of the local host, or the specified parameter, if any.
See the \fB-r\fP flag for a description how to construct matches.
Sessions marked as `dead' should be thoroughly checked and removed. 
Ask your system administrator if you are not sure. Remove sessions with the 
\fB-wipe\fP option. 
.TP 5
.B \-L
tells
.I screen
to turn on automatic output logging for the windows.
.TP 5
.BI "\-Logfile " file
By default logfile name is \*Qscreenlog.0\*Q. You can set new logfile name
with the \*Q-Logfile\*Q option.
.TP 5
.B \-m
causes
.I screen
to ignore the $STY environment variable. With \*Qscreen \-m\*U creation of
a new session is enforced, regardless whether
.I screen
is called from within another
.I screen
session or not. This flag has a special meaning in connection
with the `\-d' option:
.TP 8
.B \-d \-m
Start
.I screen
in \*Qdetached\*U mode. This creates a new session but doesn't
attach to it. This is useful for system startup scripts.
.TP 8
.B \-D \-m
This also starts screen in \*Qdetached\*U mode, but doesn't fork
a new process. The command exits if the session terminates.
.TP 5
.B \-O
selects an optimal output mode for your terminal rather than true VT100
emulation (only affects auto-margin terminals without `LP').
This can also be set in your .screenrc by specifying `OP' in a \*Qtermcap\*U
command.
.TP 5
.BI "\-p " number_or_name|-|=|+
Preselect a window. This is useful when you want to reattach to a
specific window or you want to send a command via the \*Q-X\*U
option to a specific window. As with screen's select command, \*Q-\*U
selects the blank window. As a special case for reattach, \*Q=\*U
brings up the windowlist on the blank window, while a \*Q+\*U
will create a new window. The command will not be
executed if the specified window could not be found.
.TP 5
.B \-P
Turns authentication on. By default
.I screen
allows to attach without enter the user's password.
This option changes this behavior, i.e. enables authentication.
The same effect can be done by using "auth on" command.
.TP 5
.B \-q
Suppress printing of error messages. In combination with \*Q-ls\*U the exit 
value is as follows: 9 indicates a directory without sessions. 10 
indicates a directory with running but not attachable sessions. 11 (or more) 
indicates 1 (or more) usable sessions.
In combination with \*Q-r\*U the exit value is as follows: 10 indicates that 
there is no session to resume. 12 (or more) indicates that there are 2 (or 
more) sessions to resume and you should specify which one to choose. 
In all other cases \*Q-q\*U has no effect.
.TP 5
.B \-Q
Some commands now can be queried from a remote session using this
flag, e.g. \*Qscreen \-Q windows\*U. The commands will send the
response to the stdout of the querying process. If there was an
error in the command, then the querying process will exit with
a non-zero status.

The commands that can be queried now are:
 \fBecho\fP
 \fBinfo\fP
 \fBlastmsg\fP
 \fBnumber\fP
 \fBselect\fP
 \fBtime\fP
 \fBtitle\fP
 \fBwindows\fP
.TP 5
.BR \-r " [" \fIpid.tty.host ]
.PD 0
.TP 5
.BR \-r " \fIsessionowner/[" \fIpid.tty.host ]
.PD
resumes a detached
.I screen
session.  No other options (except combinations with \fB\-d\fP/\fB\-D\fP) may
be specified, though an optional prefix of [\fIpid.\fP]\fItty.host\fP
may be needed to distinguish between multiple detached
.I screen
sessions.  The second form is used to connect to another user's screen session
which runs in multiuser mode. This indicates that screen should look for
sessions in another user's directory. This requires setuid-root.
.TP 5
.B \-R
resumes screen only when it's unambiguous which one to attach, usually
when only one
.I screen
is detached. Otherwise lists available sessions.
.B \-RR
attempts to resume the first detached
.I screen
session it finds.  If successful, all other command-line options are ignored.
If no detached session exists, starts a new session using the specified
options, just as if
.B \-R
had not been specified. The option is set by default if
.I screen
is run as a login-shell (actually screen uses \*Q\-xRR\*U in that case).
For combinations with the \fB\-d\fP/\fB\-D\fP option see there.
.TP 5
.BI "\-s " program 
sets the default shell to the program specified, instead of the value
in the environment variable $SHELL (or \*Q/bin/sh\*U if not defined).
This can also be defined through the \*Qshell\*U .screenrc command.
See also there.
.TP 5
.BI "\-S " sessionname
When creating a new session, this option can be used to specify a
meaningful name for the session. This name identifies the session for
\*Qscreen \-list\*U and \*Qscreen \-r\*U actions. It substitutes the
default [\fItty.host\fP] suffix. This name should not be longer
then 80 symbols.
.TP 5
.BI "\-t " name
sets the title (a.\|k.\|a.) for the default shell or specified program.
See also the \*Qshelltitle\*U .screenrc command.
.TP 5
.BI "\-T " term
Set the $TERM environment variable using the specified term as
opposed to the default setting of \fBscreen\fP.
.TP 5
.B \-U
Run screen in UTF-8 mode. This option tells screen that your terminal
sends and understands UTF-8 encoded characters. It also sets the default
encoding for new windows to `utf8'.
.TP 5
.B \-v
Print version number.
.TP 5
.BR \-wipe " [" \fImatch ]
does the same as \*Qscreen \-ls\*U, but removes destroyed sessions instead of
marking them as `dead'.
An unreachable session is considered dead, when its name matches either 
the name of the local host, or the explicitly given parameter, if any.
See the \fB-r\fP flag for a description how to construct matches.
.TP 5
.B \-x
Attach to a not detached
.I screen
session. (Multi display mode).
.I Screen
refuses to attach from within itself. 
But when cascading multiple screens, loops are not detected; take care.
.TP 5
.B \-X
Send the specified command to a running screen session. You may
use the \fB-S\fP option to specify the screen session if you have
several screen sessions running. You can use the \fB-d\fP or
\fB-r\fP option to tell screen to look only for attached or
detached screen sessions. Note that this command doesn't work if
the session is password protected.

.TP 5
.B \-4
Resolve hostnames only to IPv4 addresses.
.TP 5
.B \-6
Resolve hostnames only to IPv6 addresses.
.SH "DEFAULT KEY BINDINGS"
.ta 12n 26n
As mentioned, each
.I screen
command consists of a
\*QC-a\*U followed by one other character.
For your convenience, all commands that are bound to lower-case letters are
also bound to their control character counterparts (with the exception
of \*QC-a a\*U; see below), thus, \*QC-a c\*U as well as \*QC-a C-c\*U can
be used to create a window. See section \*QCUSTOMIZATION\*U for a description
of the command.
.PP
The following table shows the default key bindings. The trailing
commas in boxes with multiple keystroke entries are separators,
not part of the bindings.
.TS
tab(;);
lb l l.
_
C-a ';(select);T{
Prompt for a window name or number to switch to.
T}
_
C-a \[dq];(windowlist -b);T{
Present a list of all windows for selection.
T}
_
C-a \fIdigit\fP;(select 0-9);T{
Switch to window number 0 \- 9
T}
_
C-a -;(select -);T{
Switch to window number 0 \- 9, or to the blank window.
T}
_
C-a tab;(focus);T{
Switch the input focus to the next region.
See also \fIsplit, remove, only\fP.
T}
_
C-a C-a;(other);T{
Toggle to the window displayed previously.
Note that this binding defaults to the command character typed twice,
unless overridden.  For instance, if you use the option \*Q\fB\-e]x\fP\*U,
this command becomes \*Q]]\*U.
T}
_
C-a a  ;(meta);T{
Send the command character (C-a) to window. See \fIescape\fP command.
T}
_
C-a A;(title);T{
Allow the user to enter a name for the current window.
T}
_
T{
C-a b,
.br
C-a C-b
T};(break);T{
Send a break to window.
T}
_
C-a B;(pow_break);T{
Reopen the terminal line and send a break.
T}
_
T{
C-a c,
.br
C-a C-c
T};(screen);T{
Create a new window with a shell and switch to that window.
T}
_
C-a C\fP;(clear);T{
Clear the screen.
T}
_
T{
C-a d,
.br
C-a C-d
T};(detach);T{
Detach
.I screen
from this terminal.
T}
_
C-a D D;(pow_detach);T{
Detach and logout.
T}
_
T{
C-a f,
.br
C-a C-f\fP
T};(flow);T{
Toggle flow \fIon\fP, \fIoff\fP or \fIauto\fP.
T}
_
C-a F;(fit);T{
Resize the window to the current region size.
T}
_
C-a C-g;(vbell);T{
Toggles
.I screen's
visual bell mode.
T}
_
C-a h;(hardcopy);T{
Write a hardcopy of the current window to the file \*Qhardcopy.\fIn\fP\*U.
T}
_
C-a H;(log);T{
Begins/ends logging of the current window to the file \*Qscreenlog.\fIn\fP\*U.
T}
_
T{
C-a i,
.br
C-a C-i
T};(info);T{
Show info about this window.
T}
_
T{
C-a k,
.br
C-a C-k
T};(kill);T{
Destroy current window.
T}
_
T{
C-a l,
.br
C-a C-l
T};(redisplay);T{
Fully refresh current window.
T}
_
C-a L;(login);T{
Toggle this windows login slot. Available only if
.I screen
is configured to update the utmp database.
T}
_
T{
C-a m,
.br
C-a C-m
T};(lastmsg);T{
Repeat the last message displayed in the message line.
T}
_
C-a M;(monitor);T{
Toggles monitoring of the current window.
T}
_
T{
C-a space,
.br
C-a n,
.br
C-a C-n
T};(next);T{
Switch to the next window.
T}
_
C-a N;(number);T{
Show the number (and title) of the current window.
T}
_
T{
C-a backspace,
.br
C-a C-h,
.br
C-a p,
.br
C-a C-p
T};(prev);T{
Switch to the previous window (opposite of \fBC-a n\fP).
T}
_
T{
C-a q,
.br
C-a C-q
T};(xon);T{
Send a control-q to the current window.
T}
_
C-a Q;(only);T{
Delete all regions but the current one.
See also \fIsplit, remove, focus\fP.
T}
_
T{
C-a r,
.br
C-a C-r
T};(wrap);T{
Toggle the current window's line-wrap setting (turn the current window's
automatic margins on and off).
T}
_
T{
C-a s,
.br
C-a C-s;
T};(xoff);T{
Send a control-s to the current window.
T}
_
C-a S;(split);T{
Split the current region horizontally into two new ones.
See also \fIonly, remove, focus\fP.
T}
_
C-a v;(version);T{
Display the version and compilation date.
T}
_
C-a C-v;(digraph);T{
Enter digraph.
T}
_
T{
C-a w,
.br
C-a C-w
T};(windows);T{
Show a list of window.
T}
_
C-a W;(width);T{
Toggle 80/132 columns.
T}
_
C-a x \fRor\fP C-a C-x;(lockscreen);T{
Lock this terminal.
T}
_
C-a X\fP ;(remove);T{
Kill the current region.
See also \fIsplit, only, focus\fP.
T}
_
T{
C-a z,
.br
C-a C-z
T};(suspend);T{
Suspend
.IR screen .
Your system must support BSD-style job-control.
T}
_
C-a Z;(reset);T{
Reset the virtual terminal to its \*Qpower-on\*U values.
T}
_
C-a .;(dumptermcap);T{
Write out a \*Q.termcap\*U file.
T}
_
C-a ?;(help);T{
Show key bindings.
T}
_
C-a \e;(quit);T{
Kill all windows and terminate
.IR screen .
T}
_
C-a :;(colon);T{
Enter command line mode.
T}
_
T{
C-a [,
.br
C-a C-[\fP,
.br
C-a esc
T};(copy);T{
Enter copy/scrollback mode.
T}
_
T{
C-a C-],
.br
C-a ]
T};(paste .);T{
Write the contents of the paste buffer to the stdin queue of the
current window. 
T}
_
T{
C-a {\fP,
.br
C-a }
T};(history);T{
Copy and paste a previous (command) line.
T}
_
C-a >;(writebuf);T{
Write paste buffer to a file.
T}
_
C-a <;(readbuf);T{
Reads the screen-exchange file into the paste buffer.
T}
_
C-a =;(removebuf);T{
Removes the file used by \fBC-a <\fP and \fPC-a >\fP.
T}
_
C-a ,;(license);T{
Shows where
.I screen
comes from, where it went to and why you can use it.
T}
_
C-a _;(silence);T{
Start/stop monitoring the current window for inactivity.
T}
_
C-a |;(split -v);T{
Split the current region vertically into two new ones.
T}
_
C-a *;(displays);T{
Show a listing of all currently attached displays.
T}
_
.TE

.SH CUSTOMIZATION
The \*Qsocket directory\*U defaults either to $HOME/.screen or simply to
/tmp/screens or preferably to /usr/local/screens chosen at compile-time. If
.I screen
is installed setuid-root, then the administrator
should compile
.I screen
with an adequate (not NFS mounted) socket directory. If
.I screen
is not running setuid-root, the user can specify any mode 700 directory
in the environment variable $SCREENDIR.
.PP
When
.I screen
is invoked, it executes initialization commands from the files
\*Q/usr/local/etc/screenrc\*U and
\*Q.screenrc\*U in the user's home directory. These are the \*Qprogrammer's
defaults\*U that can be overridden in the following ways: for the
global screenrc file 
.I screen
searches for the environment variable $SYSSCREENRC (this override feature
may be disabled at compile-time). The user specific
screenrc file is searched in $SCREENRC, then $HOME/.screenrc.
The command line option \fB\-c\fP takes
precedence over the above user screenrc files.
.PP
Commands in these files are used to set options, bind functions to
keys, and to automatically establish one or more windows at the
beginning of your
.I screen
session.
Commands are listed one per line, with empty lines being ignored.
A command's arguments are separated by tabs or spaces, and may be
surrounded by single or double quotes.
A `#' turns the rest of the line into a comment, except in quotes.
Unintelligible lines are warned about and ignored.
Commands may contain references to environment variables. The 
syntax is the shell-like "$VAR " or "${VAR}". Note that this causes 
incompatibility with previous 
.I screen
versions, as now the '$'-character has to be protected with '\e' if no
variable substitution shall be performed. A string in single-quotes is also
protected from variable substitution.
.PP
Two configuration files are shipped as examples with your screen distribution:
\*Qetc/screenrc\*U and \*Qetc/etcscreenrc\*U. They contain a number of
useful examples for various commands.
.PP
Customization can also be done 'on-line'. To enter the command mode type
`C-a :'. Note that commands starting with \*Qdef\*U change default values,
while others change current settings.
.PP
The following commands are available:
.TP
.BI acladd " usernames " \fR[\fP crypted-pw \fR]\fR
.TP
.BI addacl " usernames"
.RS 0
.PP
Enable users to fully access this screen session. \fIUsernames\fP can be one
user or a comma separated list of users. This command enables to attach to the
.I screen
session and performs the equivalent of `aclchg \fIusernames\fP +rwx \&"#?\&"'.
executed. To add a user with restricted access, use the `aclchg' command below.
If an optional second parameter is supplied, it should be a crypted password
for the named user(s). `Addacl' is a synonym to `acladd'.
Multi user mode only.
.RE 
.TP
.BI aclchg " usernames permbits list"
.TP
.BI chacl " usernames permbits list"
.RS 0
.PP
Change permissions for a comma separated list of users. Permission bits are
represented as `r', `w' and `x'. Prefixing `+' grants the permission, `\-'
removes it. The third parameter is a comma separated list of commands and/or
windows (specified either by number or title). The special list `#' refers to 
all windows, `?' to all commands. if \fIusernames\fP consists of a single `*',
all known users are affected.
.PP
A command can be executed when the user has the `x' bit for it.
The user can type input to a window when he has its `w' bit set and no other
user obtains a writelock for this window. 
Other bits are currently ignored.  
To withdraw the writelock from another user in window 2:
`aclchg \fIusername\fP \-w+w 2'.
To allow read-only access to the session: `aclchg \fIusername\fP
\-w \&"#\&"'. As soon as a user's name is known to
.I screen 
he can attach to the session and (per default) has full permissions for all 
command and windows. Execution permission for the acl commands, `at' and others
should also be removed or the user may be able to regain write permission.
Rights of the special username
.B nobody
cannot be changed (see the \*Qsu\*U command).
`Chacl' is a synonym to `aclchg'.
Multi user mode only.
.RE
.TP
.BI acldel " username"
.RS 0
.PP
Remove a user from
.IR screen 's
access control list. If currently attached, all the
user's displays are detached from the session. He cannot attach again.
Multi user mode only.
.RE
.TP
.BI aclgrp " username " \fR[\fP groupname \fR]\fP
.RS 0
.PP
Creates groups of users that share common access rights. The name of the 
group is the username of the group leader. Each member of the group inherits
the permissions that are granted to the group leader. That means, if a user
fails an access check, another check is made for the group leader.
A user is removed from all groups the special value \*Qnone\*U is used for
.IR groupname .
If the second parameter is omitted all groups the user is in are listed.
.RE
.TP
\fBaclumask\fP [[ \fIusers\fP ] \fI+bits\fP | [ \fIusers\fR ] \fI-bits...\fP ]
.TP
\fBumask\fP [[ \fIusers\fP ] \fI+bits\fP | [ \fIusers\fR ] \fI-bits...\fP ]
.RS 0
.PP
This specifies the access other users have to windows that will be created by
the caller of the command.
.I Users
may be no, one or a comma separated list of known usernames. If no users are
specified, a list of all currently known users is assumed. 
.I Bits
is any combination of access control bits allowed defined with the 
\*Qaclchg\*U command. The special username \*Q?\*U predefines the access
that not yet known users will be granted to any window initially.
The special username \*Q??\*U predefines the access that not yet known 
users are granted to any command. 
Rights of the special username
.B nobody
cannot be changed (see the \*Qsu\*U command).
`Umask' is a synonym to `aclumask'.
.RE
.TP
.BI activity " message"
.RS 0
.PP
When any activity occurs in a background window that is being monitored,
.I screen
displays a notification in the message line.
The notification message can be re-defined by means of the \*Qactivity\*U
command.
Each occurrence of `%' in \fImessage\fP is replaced by
the number of the window in which activity has occurred,
and each occurrence of `^G' is replaced by the definition for bell
in your termcap (usually an audible bell).
The default message is
.PP
.nf
	'Activity in window %n'
.fi
.PP
Note that monitoring is off for all windows by default, but can be altered
by use of the \*Qmonitor\*U command (C-a M).
.RE
.TP
.BR "allpartial [ on | off ]"
.RS 0
.PP
If set to on, only the current cursor line is refreshed on window change.
This affects all windows and is useful for slow terminal lines. The 
previous setting of full/partial refresh for each window is restored
with \*Qallpartial off\*U.  This is a global flag that immediately takes effect
on all windows overriding the \*Qpartial\*U settings. It does not change the
default redraw behavior of newly created windows.
.RE
.TP
.BR "altscreen [ on | off ]"
.RS 0
.PP
If set to on, "alternate screen" support is enabled in virtual terminals,
just like in xterm.  Initial setting is `off'.
.RE
.TP
.BR "at " "[\fIidentifier\fP][" "#\fP|\fP*\fP|\fP%\fP] " "command " [ \fIargs\fP " ... ]"
.RS 0
.PP
Execute a command at other displays or windows as if it had been entered there.
\*QAt\*U changes the context (the `current window' or `current display'
setting) of the command. If the first parameter describes a 
non-unique context, the command will be executed multiple times. If the first 
parameter is of the form `\fIidentifier\fP*' then identifier is matched against
user names.  The command is executed once for each display of the selected 
user(s). If the first parameter is of the form `\fIidentifier\fP%' identifier 
is matched against displays. Displays are named after the ttys they 

src/router/zabbix/frontends/php/screen.import.php
<?php
/*
** Zabbix
** Copyright (C) 2001-2020 Zabbix SIA
**
** This program is free software; you can redistribute it and/or modify
** it under the terms of the GNU General Public License as published by
** the Free Software Foundation; either version 2 of the License, or
** (at your option) any later version.
**
** This program is distributed in the hope that it will be useful,
** but WITHOUT ANY WARRANTY; without even the implied warranty of
** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
** GNU General Public License for more details.
**
** You should have received a copy of the GNU General Public License
** along with this program; if not, write to the Free Software
** Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
**/


require_once dirname(__FILE__).'/include/config.inc.php';

$page['file'] = 'screen.import.php';
$page['title'] = _('Configuration import');
$page['type'] = detect_page_type(PAGE_TYPE_HTML);

ob_start();

require_once dirname(__FILE__).'/include/page_header.php';

$fields = [
	'rules' => [T_ZBX_STR, O_OPT, null, null, null],
	'import' => [T_ZBX_STR, O_OPT, P_SYS|P_ACT, null, null],
	'rules_preset' => [T_ZBX_STR, O_OPT, null, null, null],
	'backurl' => [T_ZBX_STR, O_OPT, null, null, null],
	'cancel' => [T_ZBX_STR, O_OPT, P_SYS, null, null]
];
check_fields($fields);


if (isset($_REQUEST['cancel'])) {
	ob_end_clean();
	redirect(CWebUser::$data['last_page']['url']);
}
ob_end_flush();

$data = [
	'rules' => [
		'groups' => ['createMissing' => false],
		'hosts' => ['updateExisting' => false, 'createMissing' => false],
		'templates' => ['updateExisting' => false, 'createMissing' => false],
		'templateScreens' => ['updateExisting' => false, 'createMissing' => false, 'deleteMissing' => false],
		'templateLinkage' => ['createMissing' => false, 'deleteMissing' => false],
		'applications' => ['createMissing' => false, 'deleteMissing' => false],
		'items' => ['updateExisting' => false, 'createMissing' => false, 'deleteMissing' => false],
		'discoveryRules' => ['updateExisting' => false, 'createMissing' => false, 'deleteMissing' => false],
		'triggers' => ['updateExisting' => false, 'createMissing' => false, 'deleteMissing' => false],
		'graphs' => ['updateExisting' => false, 'createMissing' => false, 'deleteMissing' => false],
		'httptests' => ['updateExisting' => false, 'createMissing' => false, 'deleteMissing' => false],
		'screens' => ['updateExisting' => false, 'createMissing' => false],
		'maps' => ['updateExisting' => false, 'createMissing' => false],
		'images' => ['updateExisting' => false, 'createMissing' => false],
		'mediaTypes' => ['updateExisting' => false, 'createMissing' => false],
		'valueMaps' => ['updateExisting' => false, 'createMissing' => false]
	],
	'backurl' => getRequest('backurl', '')
];

if (!CHtmlUrlValidator::validate($data['backurl'], ['allow_user_macro' => false, 'validate_uri_schemes' => true])) {
	$data['backurl'] = 'zabbix.php?action=dashboard.view';
}

// rules presets
if (hasRequest('rules_preset') && !hasRequest('rules')) {
	switch (getRequest('rules_preset')) {
		case 'host':
			$data['rules']['groups'] = ['createMissing' => true];
			$data['rules']['hosts'] = ['updateExisting' => true, 'createMissing' => true];
			$data['rules']['applications'] = ['createMissing' => true, 'deleteMissing' => false];
			$data['rules']['items'] = ['updateExisting' => true, 'createMissing' => true, 'deleteMissing' => false];
			$data['rules']['discoveryRules'] = ['updateExisting' => true, 'createMissing' => true,
				'deleteMissing' => false
			];
			$data['rules']['triggers'] = ['updateExisting' => true, 'createMissing' => true, 'deleteMissing' => false];
			$data['rules']['graphs'] = ['updateExisting' => true, 'createMissing' => true, 'deleteMissing' => false];
			$data['rules']['httptests'] = ['updateExisting' => true, 'createMissing' => true, 'deleteMissing' => false];
			$data['rules']['templateLinkage'] = ['createMissing' => true, 'deleteMissing' => false];
			$data['rules']['valueMaps'] = ['updateExisting' => false, 'createMissing' => true];

			$data['backurl'] = 'hosts.php';
			break;

		case 'template':
			$data['rules']['groups'] = ['createMissing' => true];
			$data['rules']['templates'] = ['updateExisting' => true, 'createMissing' => true];
			$data['rules']['templateScreens'] = ['updateExisting' => true, 'createMissing' => true,
				'deleteMissing' => false
			];
			$data['rules']['applications'] = ['createMissing' => true, 'deleteMissing' => false];
			$data['rules']['items'] = ['updateExisting' => true, 'createMissing' => true, 'deleteMissing' => false];
			$data['rules']['discoveryRules'] = ['updateExisting' => true, 'createMissing' => true,
				'deleteMissing' => false
			];
			$data['rules']['triggers'] = ['updateExisting' => true, 'createMissing' => true, 'deleteMissing' => false];
			$data['rules']['graphs'] = ['updateExisting' => true, 'createMissing' => true, 'deleteMissing' => false];
			$data['rules']['httptests'] = ['updateExisting' => true, 'createMissing' => true, 'deleteMissing' => false];
			$data['rules']['templateLinkage'] = ['createMissing' => true, 'deleteMissing' => false];
			$data['rules']['valueMaps'] = ['updateExisting' => false, 'createMissing' => true];

			$data['backurl'] = 'templates.php';
			break;

		case 'map':
			$data['rules']['maps'] = ['updateExisting' => true, 'createMissing' => true];

			$data['backurl'] = 'sysmaps.php';
			break;

		case 'mediatype':
			$data['rules']['mediaTypes'] = ['updateExisting' => false, 'createMissing' => true];

			$data['backurl'] = 'zabbix.php?action=mediatype.list';
			break;

		case 'screen':
			$data['rules']['screens'] = ['updateExisting' => true, 'createMissing' => true];

			$data['backurl'] = 'screenconf.php';
			break;

		case 'valuemap':
			$data['rules']['valueMaps'] = ['updateExisting' => false, 'createMissing' => true];

			$data['backurl'] = 'adm.valuemapping.php';
			break;

	}
}

if (hasRequest('rules')) {
	$requestRules = getRequest('rules', []);
	// if form was submitted with some checkboxes unchecked, those values are not submitted
	// so that we set missing values to false, existing to true
	foreach ($data['rules'] as $ruleName => $rule) {
		if (!array_key_exists($ruleName, $requestRules)) {
			$requestRules[$ruleName] = [];
		}

		foreach (['updateExisting', 'createMissing', 'deleteMissing'] as $option) {
			if (array_key_exists($option, $requestRules[$ruleName])) {
				$requestRules[$ruleName][$option] = true;
			}
			elseif (array_key_exists($option, $rule)) {
				$requestRules[$ruleName][$option] = false;
			}
		}
	}

	$data['rules'] = $requestRules;
}

if (isset($_FILES['import_file'])) {
	$result = false;

	// CUploadFile throws exceptions, so we need to catch them
	try {
		$file = new CUploadFile($_FILES['import_file']);

		$result = API::Configuration()->import([
			'format' => CImportReaderFactory::fileExt2ImportFormat($file->getExtension()),
			'source' => $file->getContent(),
			'rules' => $data['rules']
		]);
	}
	catch (Exception $e) {
		error($e->getMessage());
	}

	show_messages($result, _('Imported successfully'), _('Import failed'));
}

$view = new CView('conf.import', $data);
$view->render();
$view->show();

require_once dirname(__FILE__).'/include/page_footer.php';
